junit/src/main/java/io/cucumber/junit/FeatureRunner.java

```diff
@@ -78,23 +77,21 @@ protected void runChild(PickleRunner child, RunNotifier notifier) {
     }
 
     private void buildFeatureElementRunners(Filters filters, RunnerSupplier runnerSupplier, JUnitOptions jUnitOptions) {
-        for (PickleEvent pickleEvent : cucumberFeature.getPickles()) {
-            if (filters.matchesFilters(pickleEvent)) {
-                try {
-                    if (jUnitOptions.stepNotifications()) {
-                        PickleRunner picklePickleRunner;
-                        picklePickleRunner = withStepDescriptions(runnerSupplier, pickleEvent, jUnitOptions);
-                        children.add(picklePickleRunner);
-                    } else {
-                        PickleRunner picklePickleRunner;
-                        picklePickleRunner = withNoStepDescriptions(cucumberFeature.getName(), runnerSupplier, pickleEvent, jUnitOptions);
-                        children.add(picklePickleRunner);
-                    }
-                } catch (InitializationError e) {
-                    throw new CucumberException("Failed to create scenario runner", e);
+        cucumberFeature.getPickles().stream().filter(filters).forEach(pickleEvent -> {
+            try {
+                if (jUnitOptions.stepNotifications()) {
+                    PickleRunner picklePickleRunner;
+                    picklePickleRunner = withStepDescriptions(runnerSupplier, pickleEvent, jUnitOptions);
+                    children.add(picklePickleRunner);
+                } else {
+                    PickleRunner picklePickleRunner;
+                    picklePickleRunner = withNoStepDescriptions(cucumberFeature.getName(), runnerSupplier, pickleEvent, jUnitOptions);
+                    children.add(picklePickleRunner);
                 }
+            } catch (InitializationError e) {
+                throw new CucumberException("Failed to create scenario runner", e);
             }
-        }
+        });
```

Comment:
Changing the if block so that it becomes a map step would allow this to be rewritten to a function that returns the `children` list instead of adding into it. Then in the constructor this list can be assigned. Considering the arguments involved I would even considering in lining this in the constructor.